### ISSUE TITLE
memory alignment issue that only occurs with specific buffer sizes (image dimensions) and results in leaks

### ISSUE THREAD

**Issue by cetinsert** [emscripten-core/emscripten#14459]
In a project compiled with `emscripten` to WASM, calls to a function with specific image dimensions keep triggering

```
  memalign memset free
  memalign memset free
  memalign memset
```

for each function call. The resulting memory references then work as expected when probed BUT cannot be deallocated and leak.

---

## Single-line Reproduction and Details

https://github.com/paulo-coutinho/pdfium-lib/issues/33#issuecomment-861669272

---

I kindly ask `emscripten` experts for their help in finding how to compile the said project in a way that resolves this issue.

**Comment by cetinsert** [comment]
# `FPDFBitmap_CreateEx` call tree

- `FPDFBitmap_CreateEx` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/fpdfsdk/fpdf_view.cpp#799
  - `pdfium::MakeRetain` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/core/fxcrt/retain_ptr.h#174
  - `CFX_DIBitmap::Create` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/core/fxge/dib/cfx_dibitmap.cpp#27
    - `FX_TryAlloc` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/core/fxcrt/fx_memory.h#48
      - `Calloc` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/core/fxcrt/fx_memory.cpp#109
        - `PartitionAllocGenericFlags` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/third_party/base/allocator/partition_allocator/partition_alloc.h#394
        - `PartitionAllocatorGeneric` – https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/4542/third_party/base/allocator/partition_allocator/partition_alloc.h#517

**Comment by cetinsert** [comment]
@kripken – can increasing the `wasmMemory` block size from `64KB` to something higher also help with fragmentation?

**Comment by cetinsert** [comment]
> To verify that, I'd instrument `malloc`, `memalign`, and `free` (and also `realloc` etc. if used). Just counting the bytes allocated and freed. One way to instrument it is to compile with `WASM=0` ...

Instrumentation was easy, no need for `WASM=0`.

```js
(() => {
  const ins       = (f, n) => (...a) => { console.warn(n, f, a); return f(...a); };
  const { malloc, memalign, memset, free, realloc } = Module.asm;
  Module.asm = { ...Module.asm,
    malloc:   ins(malloc,   'malloc  '),
    memalign: ins(memalign, 'memalign'),
    memset:   ins(memset,   'memset  '),
    free:     ins(free,     'free    '),
    realloc:  ins(realloc,  'realloc ')
  };
})();
```

![image](https://user-images.githubusercontent.com/27027/122472934-29819b00-cffc-11eb-895f-8a0e56063dc6.png)

**Comment by cetinsert** [comment]
@kripken – updated the original issue with more data from the instrumentation (which does trigger repeatedly on every time as it is a new known bad case and bad cases trigger the `memalign` `⋯` pattern in a way (at a layer) that also triggers the instrumentation), please take a look at the new comment linked below to see if it reminds you of any pattern you or other developers of `emscripten` recognize.

https://github.com/paulo-coutinho/pdfium-lib/issues/33#issuecomment-863587472

**Comment by kripken** [comment]
Instrumenting `Module.asm` like that will just affect calls that go through JS. Direct calls inside the wasm would not be noticed.

If that is not a problem, then the logging does show a leak in the codebase, as there is no free for the third memalign. To investigate that you could get a stack trace for that allocation, then try to figure out where the free should be, etc.

**Comment by cetinsert** [comment]
> If it is fragmentation in fact, then the codebase might benefit from reusing buffers, using an arena, etc.

I assume there is no way to compact / vacuum the memory.

> To investigate that you could get a stack trace for that allocation, then try to figure out where the free should be, etc.

1. Is there a compile/run-time option to reduce/eliminate these `memalign` `⋯` calls to test if codebase still works?
2. What would do a `memalign(65535, ⋯)` should `65535` not be a small number that is more like `4` because WASM is `32-bit`?
3. Get a stack trace in JS or C++ or via an emscripten compilation option? (I have not compiled this module myself, have been treating it as a blackbox so far.)

**Comment by cetinsert** [comment]
> If that is not a problem, then the logging does show a leak in the codebase, as there is no free for the third memalign. To investigate that you could get a stack trace for that allocation, then try to figure out where the free should be, etc.

The original C++ codebase was checked and found to not suffer from memory leaks:

- https://bugs.chromium.org/p/pdfium/issues/detail?id=1692 (reported to Google)
- https://pdfium-review.googlesource.com/c/pdfium/+/82111/1/fpdfsdk/fpdf_view_embeddertest.cpp (Google tested)
- https://github.com/paulo-coutinho/pdfium-lib/issues/33#issuecomment-861197485 (also, self-confirmed in C++)

---

Here is how the WASM builds are being made with `emscripten` `em++` by the repository owner.

<details><summary><code>PDFium</code> build steps 👈🏻 (dockerized short version)</summary>

## How to compile (general)

1. Get the source:  
```git clone https://github.com/prsolucoes/pdfium-lib.git```  
```cd pdfium-lib```  

2. Install pip requirements:  
```pip3 install -r requirements.txt``` 

3. Get Google Depot Tools:  
```python3 make.py run build-depot-tools```  
```export PATH=$PATH:$PWD/build/depot-tools```  

Obs:
- The file **make.py** need be executed with python version 3.  
- These steps you only need make one time.  
- If you change **pdfium** git commit revision on file **modules/config.py** only execute step 4.

## How to compile for WASM

Check tutorial here: [Build for WASM](https://github.com/paulo-coutinho/pdfium-lib/blob/master/docs/BUILD_WASM.md) (docker sections pasted below)

⋮

## Docker

You can use docker to build and test on local machine before deploy.

Build the image with command:

```docker build -t pdfium-wasm -f docker/wasm/Dockerfile docker/wasm```

Test with command:

```docker run -v ${PWD}:/app -it pdfium-wasm echo "test"```

Now you can execute any command with pattern:

```docker run -v ${PWD}:/app -it pdfium-wasm [COMMAND]```

Obs: This is the recommended way to build and is used on CD server.

## Run on browser

You can test the sample using commands:

⋮

```
docker run -v ${PWD}:/app -it pdfium-wasm python3 make.py run test-wasm
python3 -m http.server --directory sample-wasm/build
```

## Run on terminal

You can test the sample using commands:

⋮


```
docker run -v ${PWD}:/app -it pdfium-wasm python3 make.py run test-wasm
docker run -v ${PWD}:/app -it pdfium-wasm node sample-wasm/build/index.js
```

## Web demo

You can test pdfium on web browser here:

https://pdfviewer.github.io/


---

### References Quoted

1. https://github.com/paulo-coutinho/pdfium-lib/blob/master/README.md
2. https://github.com/paulo-coutinho/pdfium-lib/blob/master/docs/BUILD_WASM.md

</details>

**Comment by cetinsert** [comment]
Further improved the instrumentation in https://github.com/paulo-coutinho/pdfium-lib/issues/33#issuecomment-863707842 to keep counts and added screenshots.

**Comment by kripken** [comment]
> I assume there is no way to compact / vacuum the memory.

Correct. That is a general issue with languages using linear memory. On native platforms this is also a problem, mostly on 32-bit (as on 64-bit virtual memory can be used to work around it).

You can get a stack trace programmatically by `console.log(new Error().stack)` for example. Or you can get a trace running in the debugger.

em++ may call memalign in some cases for internal reasons (like if you call `mmap`). The stack trace should help figure that out.

**Comment by cetinsert** [comment]
> You can get a stack trace programmatically by console.log(new Error().stack) for example. Or you can get a trace running in the debugger.

![image](https://user-images.githubusercontent.com/27027/122605801-e7fff700-d0b2-11eb-962c-397a41db8f90.png)

Here is the stack trace.

**Comment by kripken** [comment]
Thanks, ok, then the allocation happens here in mmap:

https://github.com/emscripten-core/emscripten/blob/7192647ca6d3add7291338b735a166fd3409edca/src/library_syscall.js#L248

and the free should happen here in munmap:

https://github.com/emscripten-core/emscripten/blob/7192647ca6d3add7291338b735a166fd3409edca/src/library_syscall.js#L302

I don't see anything obviously wrong there, and in our test suite the tests pass (and I verified we check that free is called on the right pointer).

Is it possible the codebase you are compiling does not call munmap on what it mmaps?

**Comment by cetinsert** [comment]
> Is it possible the codebase you are compiling does not call munmap on what it mmaps?

checking now ...

**Comment by cetinsert** [comment]
> Is it possible the codebase you are compiling does not call munmap on what it mmaps?

![image](https://user-images.githubusercontent.com/27027/122635706-72327480-d120-11eb-84f2-cfa8400b1ba8.png)

**Comment by cetinsert** [comment]
![image](https://user-images.githubusercontent.com/27027/122642555-fea35e00-d145-11eb-9584-fce4e32d42de.png)
Is it not safe to change this
```js
if (len === info.len)
```
to this
```js
if (len <=  info.len)
```
assuming `emscripten` bookkeeps the memory block from `info.malloc` to `info.malloc` `+` `info.len` as allocated anyway?

**Comment by sbc100** [comment]
Is the code in question trying to partial unmapping of anonymous regions?

We don't support that, and I don't think there is any easy way to fake it.    We should really be asserting in that case.  We also don't support mprotect at all.  We should assert there too.

Is the code in question doing anonymous mapping or file-backed mappings?

If its just anonymous mappings then I best thing to do is to find a way to tell the upstream code the we don't really support mmap at all.  Is there perhaps a HAS_MMAP macro that you can turn off?

**Comment by cetinsert** [comment]
@sbc100 – thank you for the prompt response.
@paulo-coutinho – can you discuss these issues with the `PDFium` group like you have done in the past?

**Comment by paulo-coutinho** [comment]
Ok, i will check it.

**Comment by cetinsert** [comment]
@kripken does that mean the latest `PDFium` cannot be compiled with `em++` without extensive source code modifications?
It does seem like you were involved with a commercial project doing just that (https://pspdfkit.com/blog/2019/webassembly-emscripten-chat-alon-zakai/) though, would you mind sharing what you already know or could learn from that company by asking with the open source community here?

**Comment by cetinsert** [comment]
Please note: https://github.com/emscripten-core/emscripten/pull/14504#issuecomment-865648164

**Comment by cetinsert** [comment]
For reference, per https://groups.google.com/g/pdfium/c/cDkppTy9mEc/m/2_4sjSdqAwAJ, `PDFium` calls both `munmap` and `mmap` only in `1` file (re-confirmed with source code search tools).

- `munmap`: https://pdfium.googlesource.com/pdfium.git/+/refs/heads/chromium/4543/third_party/base/allocator/partition_allocator/page_allocator_internals_posix.h#129
- `mmap`: https://pdfium.googlesource.com/pdfium.git/+/refs/heads/chromium/4543/third_party/base/allocator/partition_allocator/page_allocator_internals_posix.h#84

**Comment by sbc100** [comment]
Reading to code briefly it seems like the current codebase has has issues under emscirpten both before and after #14459. 
 Please tell my if my understanding not correct here:   PDFium is always trying to allocate memory with 2Mb alignment (kSuperPageSize).   Do do this it is calling `mmap` with a suggested address (arg0 which is the hint) which it aligns to 2Mb.   This hint is completely ignored by emscripten so the result will pretty much always fail to be 2Mb aligned.    When the result is not 2Mb aligned PDFium will then over-allocates by 2Mb and then try to trim the result of `mmap` by `munmap`ing the regions at the start and the end.   This munmap'ing will silently fail on emscripten prior to  #14459 and abort after #14459.

However, one key observation is that when the munmap would previously silently fail pdfium can then never free the aligned result.   See TrimMappingInternal:

```
  if (pre_slack) {                                                               
    int res = munmap(base, pre_slack);                                           
    CHECK(!res);                                                                                                               
    ret = reinterpret_cast<char*>(base) + pre_slack;                                
  }  
```

When mmunamp silently fails here PDFium adjusts the resulting pointer to form a new pointer that is now un-freeable.  This is no longer a pointer that can be passed to munmap in the future since the underlying malloc implementation is not tracking it.

I think the ultimate solution here is to patch PDFium such that under emscripten is simply calls `posix_memalign(kSuperPageSize, ...);` to get it aligned memory.

**Comment by cetinsert** [comment]
I am trying replacing 

https://pdfium.googlesource.com/pdfium.git/+/refs/heads/chromium/4543/third_party/base/allocator/partition_allocator/page_allocator_internals_posix.h#129

```cpp
void* TrimMappingInternal(void* base,
                          size_t base_length,
                          size_t trim_length,
                          PageAccessibilityConfiguration accessibility,
                          bool commit,
                          size_t pre_slack,
                          size_t post_slack) {
  void* ret = base;
  // We can resize the allocation run. Release unneeded memory before and after
  // the aligned range.
  if (pre_slack) {
    int res = munmap(base, pre_slack);
    CHECK(!res);
    ret = reinterpret_cast<char*>(base) + pre_slack;
  }
  if (post_slack) {
    int res = munmap(reinterpret_cast<char*>(ret) + trim_length, post_slack);
    CHECK(!res);
  }
  return ret;
}
```

with

https://pdfium.googlesource.com/pdfium.git/+/refs/heads/chromium/4543/third_party/base/allocator/partition_allocator/page_allocator_internals_win.h#52

```cpp
void* TrimMappingInternal(void* base,
                          size_t base_length,
                          size_t trim_length,
                          PageAccessibilityConfiguration accessibility,
                          bool commit,
                          size_t pre_slack,
                          size_t post_slack) {
  void* ret = base;
  if (pre_slack || post_slack) {
    // We cannot resize the allocation run. Free it and retry at the aligned
    // address within the freed range.
    ret = reinterpret_cast<char*>(base) + pre_slack;
    FreePages(base, base_length);
    ret = SystemAllocPages(ret, trim_length, accessibility, PageTag::kChromium,
                           commit);
  }
  return ret;
}
```

as we speak ... (so far with no success).

**Comment by sbc100** [comment]
I think that the entire `TrimMappingInternal` and `TrimMapping` should not be included in the emscripten build.  Trimming an existing mapping is not really possible with a malloc-backed mmap.     I think we should instread just use `posix_memalign` to get an allocation that is already aligned (as opposed to using the `hint` parameter to `mmap` which seems like strictly less precise way to express the intent to the underlying allocator).  

In other words, on systems which don't have real `mmap` support lets avoid `mmap` completely.

**Comment by cetinsert** [comment]
@sbc100 – That would suggest we would need a new file like `page_allocator_internals_em.h`. As my C experience is not up to speed to create one, can you or anyone else be come up with the necessary changes / a complete example? If you want compensation because this is beyond the scope of `emscripten`, please write to me at `work@pdf.ist` and we can discuss the conditions.

**Comment by sbc100** [comment]
I don't think I will have the time to add this myself, but I suggest we/you reach out to folks who work on PDFium and see if we can persuade them to add (or at least accept) a no-mmap version of that codepath.

**Comment by kripken** [comment]
@cetinsert 

Thanks for the feedback on that PR. Let's iterate there to find the right solution.

> It does seem like you were involved with a commercial project compiling PDFium with emscripten

I think that was just a friendly chat about wasm that I did on their blog. I don't know anything about their codebase.

**Comment by cetinsert** [comment]
@kripken – thank you for the friendly response! I have just learned how to compile the WASM builds without missing a step re-using a stale file and am now testing the `⋯_win.h` approach described here https://github.com/emscripten-core/emscripten/issues/14459#issuecomment-866066268 which was considered worth a try by someone from Google too (https://groups.google.com/g/pdfium/c/cDkppTy9mEc/m/EIK5_FGIAwAJ).

**Comment by cetinsert** [comment]
@sbc100 – trying `⋯_win.h` approach as described here https://github.com/emscripten-core/emscripten/issues/14459#issuecomment-866066268
seemed to help with simple cases such as

```js
_PDFium_Init(); var s = 496; for (let i = 0; i < 1; i++) FPDF.Bitmap_Destroy(FPDF.Bitmap_CreateEx(s, s, 4)); [ wasmMemory.buffer.byteLength ]
```

but it does hang when more stuff is happening such as actually rendering pages o__O.

**Comment by sbc100** [comment]
I don't think that will work.  The main reason is that the `hint` argument which is the first argument to SystemAllocPages is ignored under emscripten.    As I said, I think you are going to want a more specific, higher level, solution that avoids SystemAllocPages completely and uses something like `posix_memalign` instead.   Hopefully this can be done upstream: https://groups.google.com/g/pdfium/c/cDkppTy9mEc.

**Comment by cetinsert** [comment]
Any chance of this https://github.com/emscripten-core/emscripten/blob/7192647ca6d3add7291338b735a166fd3409edca/src/library_syscall.js#L282

getting done in `emscripten`?

**Comment by sbc100** [comment]
> Any chance of this
> 
> https://github.com/emscripten-core/emscripten/blob/7192647ca6d3add7291338b735a166fd3409edca/src/library_syscall.js#L282
> 
> getting done in `emscripten`?

Since WebAssembly doesn't have access the virtual memory subsystem its hard to see a useful/sensible way to implement this. 
 I suppose these is the possibility of writing a separate page allocator that operates on a different chunk of memory to the malloc heap?  But we have no plans to do that.  Its also something that could just as well be implemented in user space.  i.e. we don't any system primitives that make sense here.

**Comment by cetinsert** [comment]
Lei Zhang from `PDFium` has signaled willingness in accepting a WASM-only memory management patch: https://groups.google.com/g/pdfium/c/cDkppTy9mEc/m/k976It-QAwAJ

@sbc100 @kripken – the best I can do (with my active knowledge of `C/C++`) is extend this approach https://github.com/emscripten-core/emscripten/issues/14459#issuecomment-866066268 by doing a full translation of `⋯_win.h` and use that for WASM. If you look at the `⋯win.h` linked in that comment, do you still see problems remaining because of `hint` or is Windows behaving like WASM in [`SystemAllocPages`](https://pdfium.googlesource.com/pdfium.git/+/refs/heads/chromium/4543/third_party/base/allocator/partition_allocator/page_allocator_internals_win.h#38) too?

**Comment by sbc100** [comment]
I'm talking to some folks who work on PDFium about getting a upstream patch.

**Comment by cetinsert** [comment]
@sbc100 – hello!

Is anyone making any progress on this? 
Can you provide us (@paulo-coutinho, @cetinsert) links and contacts where / with whom we can follow up?

**Comment by sbc100** [comment]
Sorry I haven't had any time to look into it.   We have approval to make an emscripten-specific version of the page allocator in PDFium but I don't think I will have time to do it myself.

**Comment by paulo-coutinho** [comment]
@cetinsert 

Do you have knowledge to do it?

**Comment by cetinsert** [comment]
@paulo-coutinho – I have limited experience with memory allocators. The fanciest thing I have ever done is `LD_LIBRARY_PATH=` a different `malloc` library to replace the `glibc` one which had been the cause of incessant `segfault`s in a Linux installation that was half-`apt upgrade`-d; no changes in legacy code, fix working like a charm: no `segfault`s ever happening again, no memory corruption either. It would speed things up if we connect with the right people as any output from me will likely take weeks at best. Perhaps @sbc100 or @kripken could briefly sketch out in pseudo-code or plaintext what needs to be replaced and how so we can give it a try!

**Comment by cetinsert** [comment]
@paulo-coutinho – `ffmpeg` was recently ported to WebAssembly (a few months ago); perhaps their community might have had to deal with / get past similar issues

**Comment by paulo-coutinho** [comment]
Hi,

Any solution for this?

Thanks.

**Comment by paulocoutinhox** [comment]
Any news on this?
